measured.go

```diff
@@ -1,310 +1,148 @@
-			MaxOut:   out,
-			LastOut:  out,
-			TotalOut: out,
-		}
-		m.traffic[id] = t
-		return
-	}
-
-	// Add to existing ID
-	if in < t.MinIn {
-		t.MinIn = in
+func (c *conn) Write(b []byte) (int, error) {
+	c.sent.begin(mtime.Now)
+	n, err := c.Conn.Write(b)
+	c.sent.advance(n, mtime.Now())
+	if err != nil && !isTimeout(err) {
+		c.storeError(err)
 	}
-	if in > t.MaxIn {
-		t.MaxIn = in
-	}
-	t.LastIn = in
-	t.TotalIn += in
-	if out < t.MinOut {
-		t.MinOut = out
-	}
-	if out > t.MaxOut {
-		t.MaxOut = out
-	}
-	t.LastOut = out
-	t.TotalOut += out
+	return n, err
 }
 
-func (m *Measured) reportTraffic() {
-	m.mutex.Lock()
-	currentTraffic := m.traffic
-	m.traffic = make(map[string]*TrafficTracker)
-	m.mutex.Unlock()
-	log.Debugf("Reporting %d traffic entries", len(currentTraffic))
-	for _, r := range m.reporters {
-		if err := r.ReportTraffic(currentTraffic); err != nil {
-			log.Errorf("Failed to report traffic data to %v: %v", reflect.TypeOf(r), err)
-		}
+func (c *conn) Read(b []byte) (int, error) {
+	c.recv.begin(mtime.Now)
+	n, err := c.Conn.Read(b)
+	c.recv.advance(n, mtime.Now())
+	if err != nil && !isTimeout(err) && err != io.EOF {
+		c.storeError(err)
 	}
+	return n, err
 }
 
-// Conn wraps any net.Conn to add statistics
-type Conn struct {
-	net.Conn
-	// arbitrary string to identify this connection, defaults to remote address
-	ID string
-	// total bytes read from this connection
-	BytesIn uint64
-	// total bytes wrote to this connection
-	BytesOut                  uint64
-	m                         *Measured
-	interval                  time.Duration
-	elapsedSinceLastSubmitted func() time.Duration
-	submitMutex               sync.Mutex
-}
-
-func (m *Measured) newConn(c net.Conn, interval time.Duration) net.Conn {
-	ra := c.RemoteAddr()
-	if ra == nil {
-		panic("nil remote address is not allowed")
+func (c *conn) Close() error {
+	if atomic.CompareAndSwapInt32(&c.closed, 0, 1) {
+		err := c.Conn.Close()
+		// Wait for tracking to finish
+		<-c.trackingFinished
+		return err
 	}
-	return &Conn{Conn: c, ID: ra.String(), m: m, interval: interval}
-}
-
-// Read() implements the function from net.Conn
-func (mc *Conn) Read(b []byte) (n int, err error) {
-	n, err = mc.Conn.Read(b)
-	atomic.AddUint64(&mc.BytesIn, uint64(n))
-	mc.submitTrafficIfNecessary()
-	return
-}
-
-// Write() implements the function from net.Conn
-func (mc *Conn) Write(b []byte) (n int, err error) {
-	n, err = mc.Conn.Write(b)
-	atomic.AddUint64(&mc.BytesOut, uint64(n))
-	mc.submitTrafficIfNecessary()
-	return
-}
-
-// Close implements the function from net.Conn
-func (mc *Conn) Close() (err error) {
-	err = mc.Conn.Close()
-	mc.submitTraffic()
-	return
+	return nil
 }
 
-func (mc *Conn) submitTrafficIfNecessary() {
-	needToSubmit := false
-	mc.submitMutex.Lock()
-	if mc.elapsedSinceLastSubmitted == nil || mc.elapsedSinceLastSubmitted() > mc.interval {
-		needToSubmit = true
-		mc.elapsedSinceLastSubmitted = mtime.Stopwatch()
+func (c *conn) storeError(err error) {
```

Comment:
Maybe also check for a `nil` `err` and `isTimeout` here instead of in the callers @oxtoacart? Not sure how it was in the original code, but just a little more DRY.